packages/navigation/src/UIGridList/UIMasonryList.tsx

```diff
@@ -0,0 +1,498 @@
+function binarySearch(coord: number, indexes: { y: number }[]) {
+    if (indexes.length === 0) {
+        return 0;
+    }
+
+    let left = 0;
+    let right = indexes.length;
+    // eslint-disable-next-line no-constant-condition
+    while (true) {
+        const middle = Math.trunc((right + left) / 2);
+        const item = indexes[middle];
+        if (item.y >= coord) {
+            right = middle;
+        } else {
+            left = middle;
+        }
+
+        if (right - left <= 1) {
+            if (indexes[left].y >= coord) {
+                return left;
+            }
+            return right;
+        }
+    }
+}
+
+type UIMasonryListCellRef = {
+    show: () => void;
+    hide: () => void;
+};
+
+type UIMasonryListCellProps<Item = any> = {
+    item: MasonryItem<Item>;
+    renderItem: (item: MasonryItem<Item>) => React.ReactNode;
+    width: number;
+    layout: LayoutCell;
+};
+
+const UIMasonryListCell = React.memo(
+    React.forwardRef<UIMasonryListCellRef, UIMasonryListCellProps>(function UIMasonryListCell(
+        { item, renderItem, width, layout }: UIMasonryListCellProps,
+        ref,
+    ) {
+        const [isVisible, setIsVisible] = React.useState(false);
+        React.useImperativeHandle(ref, () => ({
+            show() {
+                setIsVisible(true);
+            },
+            hide() {
+                setIsVisible(false);
+            },
+        }));
+
+        if (!isVisible) {
+            return null;
+        }
+
+        return (
+            <View
+                style={{
+                    position: 'absolute',
+                    top: layout.y,
+                    left: layout.x,
+                    width,
+                    height: layout.height,
+                }}
+            >
+                {renderItem(item)}
+            </View>
+        );
+    }),
+);
+
+type MasonryCellsRefs = Record<string, React.RefObject<UIMasonryListCellRef>>;
+
+function getMasonryCellsRefs<Item>(data: MasonryItem<Item>[]) {
+    return data.reduce<MasonryCellsRefs>((acc, { key }) => {
+        if (acc[key] != null) {
+            return acc;
+        }
+        acc[key] = React.createRef<UIMasonryListCellRef>();
+        return acc;
```

Comment:
Not sure if there is any benefit, just a shorter record (maybe it's simpler to read though).
```suggestion
        if (acc[key] == null) {
            acc[key] = React.createRef<UIMasonryListCellRef>();
        }
        return acc;
```